File: src/im/mark_projection.py
"""Deterministic occurrence projection through full-snapshot revisions."""

from collections.abc import Iterable
from difflib import SequenceMatcher

from im.schema.actions import Span
from im.schema.events import SnapshotEvent
from im.schema.textspan import py_index, utf16_len, utf16_slice

type RevisionOpcode = tuple[str, int, int, int, int]


def _revision_opcodes(previous: str, current: str) -> tuple[RevisionOpcode, ...]:
    """Return deterministic disjoint edit alignment with repeated-text heuristics disabled."""
    return tuple(SequenceMatcher(None, previous, current, autojunk=False).get_opcodes())


def _project_python_range(
    previous: str,
    current: str,
    start: int,
    end: int,
) -> tuple[int, int] | None:
    block = _equal_block_for_range(previous, current, start, end)
    if block is not None:
        previous_start, _previous_end, current_start, _current_end = block
        return current_start + start - previous_start, current_start + end - previous_start
    return None


def _equal_block_for_range(
    previous: str,
    current: str,
    start: int,
    end: int,
) -> tuple[int, int, int, int] | None:
    for tag, previous_start, previous_end, current_start, current_end in _revision_opcodes(
        previous, current
    ):
        if tag == "equal" and start >= previous_start and end <= previous_end:
            return previous_start, previous_end, current_start, current_end
    return None


def _occurrences(text: str, fragment: str) -> tuple[int, ...]:
    matches: list[int] = []
    start = 0
    while True:
        index = text.find(fragment, start)
        if index < 0:
            return tuple(matches)
        matches.append(index)
        start = index + 1


def _project_unique_python_range(
    previous: str,
    current: str,
    start: int,
    end: int,
) -> tuple[int, int] | None:
    """Map only through an unchanged context block unique on both sides.

    SequenceMatcher supplies a deterministic alignment, but repeated equal blocks can admit another
    equally plausible occurrence identity. Requiring the containing maximal equal block to occur
    exactly once in each snapshot turns the alignment into a conservative identity proof.
    """
    block = _equal_block_for_range(previous, current, start, end)
    if block is None:
        return None
    previous_start, previous_end, current_start, _current_end = block
    context = previous[previous_start:previous_end]
    if len(_occurrences(previous, context)) != 1 or len(_occurrences(current, context)) != 1:
        return None
    return current_start + start - previous_start, current_start + end - previous_start


def project_span(
    source_span: Span,
    snapshots: Iterable[SnapshotEvent],
) -> Span | None:
    """Carry one source occurrence only through regions unchanged by every revision.

    Prefix/suffix preservation is deliberately conservative. A target touched by a revision's
    changed middle has ambiguous identity and is dropped instead of being guessed from matching
    text elsewhere.
    """
    ordered = tuple(snapshots)
    source_index = next(
        (index for index, snapshot in enumerate(ordered) if snapshot.id == source_span.event_id),
        None,
    )
    if source_index is None:
        return None

    source = ordered[source_index]
    try:
        if utf16_slice(
            source.payload.text,
            source_span.start_utf16,
            source_span.end_utf16,
        ) != source_span.text:
            return None
        start = py_index(source.payload.text, source_span.start_utf16)
        end = py_index(source.payload.text, source_span.end_utf16)
    except (TypeError, ValueError):
        return None

    previous_text = source.payload.text
    latest_event_id = source.id
    for snapshot in ordered[source_index + 1 :]:
        projected = _project_unique_python_range(
            previous_text,
            snapshot.payload.text,
            start,
            end,
        )
        if projected is None:
            return None
        start, end = projected
        previous_text = snapshot.payload.text
        latest_event_id = snapshot.id

    if previous_text[start:end] != source_span.text:
        return None
    return Span(
        event_id=latest_event_id,
        start_utf16=utf16_len(previous_text[:start]),
        end_utf16=utf16_len(previous_text[:end]),
        text=source_span.text,
    )


def _matching_occurrences(
    text: str,
    needle: str,
    candidate_start: int,
    candidate_end: int,
) -> tuple[tuple[int, int], ...]:
    """Return equal-text occurrences local to one candidate's aligned edit interval."""
    matches: list[tuple[int, int]] = []
    start = 0
    while True:
        index = text.find(needle, start)
        if index < 0:
            break
        end = index + len(needle)
        overlaps_candidate = (
            index < candidate_end and end > candidate_start
            if candidate_start < candidate_end
            else index < candidate_start < end
        )
        if overlaps_candidate:
            matches.append((index, end))
        start = index + 1
    return tuple(matches)


def project_ambiguous_mark_targets(
    target: Span,
    snapshots: Iterable[SnapshotEvent],
) -> tuple[Span, ...]:
    """Map only the occurrence set that can descend from a touched marked target.

    Unchanged prefix/suffix regions preserve exact occurrence identity. When a revision touches a
    candidate, only equal-text occurrences overlapping that revision's changed window remain
    candidates. Equal text elsewhere is never admitted, and an empty candidate set is terminal.
    """
    ordered = tuple(snapshots)
    source_index = next(
        (index for index, snapshot in enumerate(ordered) if snapshot.id == target.event_id),
        None,
    )
    if source_index is None:
        return ()

    source = ordered[source_index]
    try:
        if utf16_slice(
            source.payload.text,
            target.start_utf16,
            target.end_utf16,
        ) != target.text:
            return ()
        candidates = {
            (
                py_index(source.payload.text, target.start_utf16),
                py_index(source.payload.text, target.end_utf16),
            )
        }
    except (TypeError, ValueError):
        return ()

    previous_text = source.payload.text
    touched = False
    latest_event_id = source.id
    for snapshot in ordered[source_index + 1 :]:
        current_text = snapshot.payload.text
        opcodes = _revision_opcodes(previous_text, current_text)
        mapped: set[tuple[int, int]] = set()
        for start, end in candidates:
            exact = _project_unique_python_range(previous_text, current_text, start, end)
            if exact is not None:
                mapped.add(exact)
                continue
            touched = True
            repeated_block = _equal_block_for_range(previous_text, current_text, start, end)
            if repeated_block is not None:
                previous_start, previous_end, _current_start, _current_end = repeated_block
                context = previous_text[previous_start:previous_end]
                relative_start = start - previous_start
                relative_end = end - previous_start
                for context_start in _occurrences(current_text, context):
                    candidate = (
                        context_start + relative_start,
                        context_start + relative_end,
                    )
                    if current_text[candidate[0] : candidate[1]] == target.text:
                        mapped.add(candidate)
                continue
            affected = [
                (current_start, current_end)
                for _tag, previous_start, previous_end, current_start, current_end in opcodes
                if (
                    max(start, previous_start) < min(end, previous_end)
                    or (previous_start == previous_end and start < previous_start < end)
                )
            ]
            if not affected:
                continue
            candidate_start = min(item[0] for item in affected)
            candidate_end = max(item[1] for item in affected)
            replacements = _matching_occurrences(
                current_text,
                target.text,
                candidate_start,
                candidate_end,
            )
            # A coarse replacement containing multiple equal strings cannot identify which one
            # descends from this occurrence. Dropping is safer than suppressing unrelated marks.
            if len(replacements) == 1:
                mapped.add(replacements[0])
        if not mapped:
            return ()
        candidates = mapped
        previous_text = current_text
        latest_event_id = snapshot.id

    if not touched:
        return ()
    return tuple(
        Span(
            event_id=latest_event_id,
            start_utf16=utf16_len(previous_text[:start]),
            end_utf16=utf16_len(previous_text[:end]),
            text=target.text,
        )
        for start, end in sorted(candidates)
        if previous_text[start:end] == target.text
    )

File: client/src/review/mark-projection.ts
/** Exact TypeScript port of src/im/mark_projection.py (autojunk=False, no junk). */

import type { AppliedMark } from "./reducer";

type Range = [number, number];
type Match = [number, number, number];
type Opcode = ["equal" | "replace" | "delete" | "insert", number, number, number, number];

export type MarkProjection = {
  applied: AppliedMark | null;
  ambiguous: AppliedMark[];
};

function codepoints(text: string): string[] {
  return Array.from(text);
}

function utf16ToCodepoint(chars: string[], offset: number): number | null {
  if (!Number.isInteger(offset) || offset < 0) return null;
  let utf16 = 0;
  for (let i = 0; i <= chars.length; i++) {
    if (utf16 === offset) return i;
    if (i === chars.length) break;
    utf16 += chars[i].length;
    if (utf16 > offset) return null;
  }
  return null;
}

function codepointToUtf16(chars: string[], offset: number): number {
  let utf16 = 0;
  for (let i = 0; i < offset; i++) utf16 += chars[i].length;
  return utf16;
}

function sliceEquals(a: string[], start: number, end: number, b: string[]): boolean {
  if (end - start !== b.length) return false;
  for (let i = 0; i < b.length; i++) if (a[start + i] !== b[i]) return false;
  return true;
}

function occurrences(haystack: string[], needle: string[]): number[] {
  if (needle.length === 0) return [];
  const found: number[] = [];
  for (let start = 0; start <= haystack.length - needle.length; start++) {
    if (sliceEquals(haystack, start, start + needle.length, needle)) found.push(start);
  }
  return found;
}

/** Python difflib.SequenceMatcher matching blocks for autojunk=False and no junk predicate. */
function matchingBlocks(a: string[], b: string[]): Match[] {
  const b2j = new Map<string, number[]>();
  for (let j = 0; j < b.length; j++) {
    const positions = b2j.get(b[j]);
    if (positions) positions.push(j);
    else b2j.set(b[j], [j]);
  }

  const longest = (alo: number, ahi: number, blo: number, bhi: number): Match => {
    let besti = alo;
    let bestj = blo;
    let bestsize = 0;
    let j2len = new Map<number, number>();
    for (let i = alo; i < ahi; i++) {
      const next = new Map<number, number>();
      for (const j of b2j.get(a[i]) ?? []) {
        if (j < blo) continue;
        if (j >= bhi) break;
        const size = (j2len.get(j - 1) ?? 0) + 1;
        next.set(j, size);
        if (size > bestsize) {
          besti = i - size + 1;
          bestj = j - size + 1;
          bestsize = size;
        }
      }
      j2len = next;
    }
    while (besti > alo && bestj > blo && a[besti - 1] === b[bestj - 1]) {
      besti--;
      bestj--;
      bestsize++;
    }
    while (
      besti + bestsize < ahi &&
      bestj + bestsize < bhi &&
      a[besti + bestsize] === b[bestj + bestsize]
    ) {
      bestsize++;
    }
    return [besti, bestj, bestsize];
  };

  const queue: [number, number, number, number][] = [[0, a.length, 0, b.length]];
  const matches: Match[] = [];
  while (queue.length > 0) {
    const [alo, ahi, blo, bhi] = queue.pop()!;
    const [i, j, size] = longest(alo, ahi, blo, bhi);
    if (size === 0) continue;
    matches.push([i, j, size]);
    if (alo < i && blo < j) queue.push([alo, i, blo, j]);
    if (i + size < ahi && j + size < bhi) {
      queue.push([i + size, ahi, j + size, bhi]);
    }
  }
  matches.sort((x, y) => x[0] - y[0] || x[1] - y[1]);
  const collapsed: Match[] = [];
  for (const match of matches) {
    const prior = collapsed.at(-1);
    if (prior && prior[0] + prior[2] === match[0] && prior[1] + prior[2] === match[1]) {
      prior[2] += match[2];
    } else {
      collapsed.push([...match]);
    }
  }
  collapsed.push([a.length, b.length, 0]);
  return collapsed;
}

function opcodes(a: string[], b: string[]): Opcode[] {
  const out: Opcode[] = [];
  let i = 0;
  let j = 0;
  for (const [ai, bj, size] of matchingBlocks(a, b)) {
    if (i < ai || j < bj) {
      const tag = i < ai && j < bj ? "replace" : i < ai ? "delete" : "insert";
      out.push([tag, i, ai, j, bj]);
    }
    if (size > 0) out.push(["equal", ai, ai + size, bj, bj + size]);
    i = ai + size;
    j = bj + size;
  }
  return out;
}

function equalBlockForRange(
  previous: string[],
  current: string[],
  start: number,
  end: number,
): [number, number, number, number] | null {
  for (const [tag, previousStart, previousEnd, currentStart, currentEnd] of opcodes(
    previous,
    current,
  )) {
    if (tag === "equal" && start >= previousStart && end <= previousEnd) {
      return [previousStart, previousEnd, currentStart, currentEnd];
    }
  }
  return null;
}

function uniqueProjection(
  previous: string[],
  current: string[],
  start: number,
  end: number,
): Range | null {
  const block = equalBlockForRange(previous, current, start, end);
  if (!block) return null;
  const [previousStart, previousEnd, currentStart] = block;
  const context = previous.slice(previousStart, previousEnd);
  if (occurrences(previous, context).length !== 1 || occurrences(current, context).length !== 1) {
    return null;
  }
  return [currentStart + start - previousStart, currentStart + end - previousStart];
}

function matchingOccurrences(
  text: string[],
  needle: string[],
  candidateStart: number,
  candidateEnd: number,
): Range[] {
  return occurrences(text, needle)
    .map((start): Range => [start, start + needle.length])
    .filter(([start, end]) =>
      candidateStart < candidateEnd
        ? start < candidateEnd && end > candidateStart
        : start < candidateStart && candidateStart < end,
    );
}

function ambiguousProjection(
  previous: string[],
  current: string[],
  start: number,
  end: number,
  target: string[],
): Range[] {
  const exact = uniqueProjection(previous, current, start, end);
  if (exact) return [];

  const repeatedBlock = equalBlockForRange(previous, current, start, end);
  if (repeatedBlock) {
    const [previousStart, previousEnd] = repeatedBlock;
    const context = previous.slice(previousStart, previousEnd);
    const relativeStart = start - previousStart;
    const relativeEnd = end - previousStart;
    return occurrences(current, context)
      .map((contextStart): Range => [
        contextStart + relativeStart,
        contextStart + relativeEnd,
      ])
      .filter(([candidateStart, candidateEnd]) =>
        sliceEquals(current, candidateStart, candidateEnd, target),
      );
  }

  const affected = opcodes(previous, current)
    .filter(
      ([, previousStart, previousEnd]) =>
        Math.max(start, previousStart) < Math.min(end, previousEnd) ||
        (previousStart === previousEnd && start < previousStart && previousStart < end),
    )
    .map(([, , , currentStart, currentEnd]): Range => [currentStart, currentEnd]);
  if (affected.length === 0) return [];
  const candidateStart = Math.min(...affected.map((item) => item[0]));
  const candidateEnd = Math.max(...affected.map((item) => item[1]));
  const replacements = matchingOccurrences(current, target, candidateStart, candidateEnd);
  return replacements.length === 1 ? replacements : [];
}

/** Project one current mark through one subsequent snapshot using canonical occurrence identity. */
export function projectAppliedMark(
  mark: AppliedMark,
  previousText: string,
  snapshotEventId: string,
  currentText: string,
): MarkProjection {
  const previous = codepoints(previousText);
  const current = codepoints(currentText);
  const target = codepoints(mark.targetText);
  const start = utf16ToCodepoint(previous, mark.targetStart);
  const end = utf16ToCodepoint(previous, mark.targetEnd);
  if (start === null || end === null || !sliceEquals(previous, start, end, target)) {
    return { applied: null, ambiguous: [] };
  }

  const make = ([candidateStart, candidateEnd]: Range): AppliedMark => ({
    ...mark,
    targetEventId: snapshotEventId,
    targetStart: codepointToUtf16(current, candidateStart),
    targetEnd: codepointToUtf16(current, candidateEnd),
  });
  const exact = uniqueProjection(previous, current, start, end);
  if (exact) return { applied: make(exact), ambiguous: [] };
  return {
    applied: null,
    ambiguous: ambiguousProjection(previous, current, start, end, target).map(make),
  };
}
